src/server/rdb_load.cc

```diff
@@ -1644,6 +1644,12 @@ RdbLoader::RdbLoader(ScriptMgr* script_mgr) : script_mgr_(script_mgr) {
 }
 
 RdbLoader::~RdbLoader() {
+  while (true) {
+    Item* item = item_queue_.Pop();
+    if (item == nullptr)
+      break;
+    delete item;
+  }
```

Comment:
I just thought that you can get a memory leak here but found an even more serious issue:

- The loader parses items and fills the shard set up to maximum with tasks
- An error occurs and stop_early_ is set, the loader returns
- The receiving fiber (for example FullSyncFb in replica) reports the error and exits, destructor runs
- You shard set has still pending tasks and they are executed, but the loader is already deleted

The solution would be removing the fast path inside Load() so that it waits for the shard set

What is more, its impossible to stop pending tasks on cancellation currently, so even after cleaning the db they'll continue inserting

But this is likely out of scope for this PR

Reply:
This scenario should be catchable with pytests on `test_crash_master`, probably with higher load

Reply:
You are right. Write it down somewhere.